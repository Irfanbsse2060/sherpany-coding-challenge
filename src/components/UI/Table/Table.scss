@import '../../../styles/variables';

$athens-gray: #f4f5f7;

$actions-column-width: 120px;
$checkbox-column-width: 60px;
$column-padding: 10px;
$icon-size: 30px;

.custom-table {
  border-collapse: collapse;
  width: 100%;

  &__header {
    &__column {
      text-align: left;
    }
  }

  &__row:nth-child(even) {
    background-color: #dddddd;
  }

  &__row {

    &--header-column, &--body-column {
      border: 1px solid #dddddd;
      text-align: left;
      padding: 8px;
    }
  }
}


