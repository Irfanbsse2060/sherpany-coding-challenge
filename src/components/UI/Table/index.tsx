import React from 'react'
import './Table.scss'

interface Header {
    name: string,
    label: string,
    customComponent?: (value: string) => React.ReactElement
}

interface CommentListProps {
    data: any[],
    header: Header[],
    testId?: string
}

export default function Table(props: CommentListProps) {

    const {data, header, testId} = props

    return (
        <table className='custom-table'>
            <tr className='custom-table__row'>
                {
                    header.map((item, index) => {
                        return <th className='custom-table__row--header-column' key={index}>{item.label}</th>
                    })
                }
            </tr>
            {
                data.map((item, index) => {
                    return (
                        <tr key={index} className='custom-table__row' data-testid='custom-table-data-rows'>
                            {
                                header.map((head) => {
                                    return (<td className='custom-table__row--body-column'>
                                        {head.customComponent? head.customComponent(item[head.name]): item[head.name]}
                                    </td>)
                                })
                            }
                        </tr>
                    )
                })
            }
        </table>
    )

}
