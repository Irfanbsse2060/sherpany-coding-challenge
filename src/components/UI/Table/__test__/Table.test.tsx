import React from 'react';
import { render } from '@testing-library/react';
import userfixture from '../../../../utils/fixtures/userfixture'
import Table from '../index'

describe('Comment List', ()=>{

    test('should be able to render table', () => {
        const headers = [
            {
                name: "thumbnail",
                label: "Profile Image",
                customComponent: (value: string)=> (<div data-testid='custom-component'>custom component {value}</div>)
            },
            {
                name: "firstName",
                label: "First Name"
            },
            {
                name: "lastName",
                label: "Last Name"
            },
            {
                name: "email",
                label: "Email"
            }
        ]
        const { container, getAllByTestId } = render(<Table  header={headers} data={userfixture}/>);
        expect(getAllByTestId('custom-table-data-rows')).toHaveLength(4)
        expect(container).toMatchSnapshot()
    })


})
