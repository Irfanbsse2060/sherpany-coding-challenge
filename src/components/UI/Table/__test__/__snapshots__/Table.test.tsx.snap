// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment List should be able to render table 1`] = `
<div>
  <table
    class="custom-table"
  >
    <tr
      class="custom-table__row"
    >
      <th
        class="custom-table__row--header-column"
      >
        Profile Image
      </th>
      <th
        class="custom-table__row--header-column"
      >
        First Name
      </th>
      <th
        class="custom-table__row--header-column"
      >
        Last Name
      </th>
      <th
        class="custom-table__row--header-column"
      >
        Email
      </th>
    </tr>
    <tr
      class="custom-table__row"
      data-testid="custom-table-data-rows"
    >
      <td
        class="custom-table__row--body-column"
      >
        <div
          data-testid="custom-component"
        >
          custom component 
        </div>
      </td>
      <td
        class="custom-table__row--body-column"
      />
      <td
        class="custom-table__row--body-column"
      />
      <td
        class="custom-table__row--body-column"
      >
        lachlan.roberts@example.com
      </td>
    </tr>
    <tr
      class="custom-table__row"
      data-testid="custom-table-data-rows"
    >
      <td
        class="custom-table__row--body-column"
      >
        <div
          data-testid="custom-component"
        >
          custom component 
        </div>
      </td>
      <td
        class="custom-table__row--body-column"
      />
      <td
        class="custom-table__row--body-column"
      />
      <td
        class="custom-table__row--body-column"
      >
        leona.campbell@example.com
      </td>
    </tr>
    <tr
      class="custom-table__row"
      data-testid="custom-table-data-rows"
    >
      <td
        class="custom-table__row--body-column"
      >
        <div
          data-testid="custom-component"
        >
          custom component 
        </div>
      </td>
      <td
        class="custom-table__row--body-column"
      />
      <td
        class="custom-table__row--body-column"
      />
      <td
        class="custom-table__row--body-column"
      >
        emilie.nielsen@example.com
      </td>
    </tr>
    <tr
      class="custom-table__row"
      data-testid="custom-table-data-rows"
    >
      <td
        class="custom-table__row--body-column"
      >
        <div
          data-testid="custom-component"
        >
          custom component 
        </div>
      </td>
      <td
        class="custom-table__row--body-column"
      />
      <td
        class="custom-table__row--body-column"
      />
      <td
        class="custom-table__row--body-column"
      >
        ester.woll@example.com
      </td>
    </tr>
  </table>
</div>
`;
